### PR TITLE
feat: add guild territory defense level sorting

### PR DESCRIPTION
**Pull request by ryanzhoudev**
Adds a button to only show a certain difficulty of guild territory defense on the guild map. Currently uses the search icon, though this can be changed to anything else.

This PR also does some refactoring and changes how MapButtons are handled. Tested in game; seems to work fine.

### CODE REVIEW

File: src/main/java/com/wynntils/modules/map/overlays/objects/MapButton.java [hunk 1]
```diff
@@ -73,7 +75,7 @@ public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
     }
 
     public Boolean isEnabled() {
-        return isEnabled.apply(null);
+        return Boolean.parseBoolean(String.valueOf(state.apply(null)));
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/map/overlays/objects/MapButton.java]: This looks a bit strange, is there any better way to do this?

File: src/main/java/com/wynntils/modules/map/overlays/objects/MapTerritory.java [hunk 2]
```diff
@@ -106,6 +106,24 @@ public TerritoryProfile getTerritory() {
...
+        } else if (def.contains("Medium")) {
+            return 3;
+        } else if (def.contains("High")) {
+            return 4;
+        } else if (def.contains("Very High")) {
+            return 5;
+        } else {
+            return -1;
+        }
+    }
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/map/overlays/objects/MapTerritory.java]: Why does equals not work here? Perhaps strip color codes? Or is there an other reason (if so maybe state that in the comment)?

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java [hunk 3]
```diff
@@ -42,6 +42,9 @@
     private boolean showTradeRoutes = true;
     private boolean territoryManageShortcut = true;
 
+    MapButton cycleButton;
+    private int territoryDefenseFilter = 0; // 0=off; 1=verylow; 2=low; 3=med; 4=high; 5=veryhigh
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: Wouldn't an enum be more useful here?

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java [hunk 4]
```diff
@@ -57,37 +60,72 @@ public void initGui() {
...
+                return "§aLow";
+            case 3:
+                return "§eMedium";
+            case 4:
+                return "§cHigh";
+            case 5:
+                return "§cVery High";
+            default:
+                return "§7Off";
+        }
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: I personally prefer to use `TextFormatting` in cases like this, but it's up to you.

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java [hunk 5]
```diff
@@ -138,8 +176,19 @@ protected void drawIcons(int mouseX, int mouseY, float partialTicks) {
...
         if (showTradeRoutes) generateTradeRoutes();
 
-        territories.values().forEach(c -> c.drawScreen(mouseX, mouseY, partialTicks, showTerritory, resourceColors, !showOwners, showOwners));
-        territories.values().forEach(c -> c.postDraw(mouseX, mouseY, partialTicks, width, height));
+        HashMap<String, MapTerritory> renderableTerritories = new HashMap<>();
+        if (territoryDefenseFilter != 0) { // If active
+            for (Map.Entry<String, MapTerritory> territory : territories.entrySet()) {
+                if (territory.getValue().getDefenses() == territoryDefenseFilter) {
+                    renderableTerritories.put(territory.getKey(), territory.getValue());
+                }
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: From what I see here, there is no way to filter like `Medium or higher` territories?

Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: Also, maybe try doing something about nesting.

**Review by kristofbolyai — COMMENTED** [review]
Looks good mostly, noted some things you should consider

**Comment by kristofbolyai** [comment]
Also, just to note, this PR `most likely` won't be merged in this release version. (Not anything related to the code, it looks awesome, we just want to follow our time schedule and don't want to merge anything without a few days of testing in CE)

File: src/main/java/com/wynntils/modules/map/overlays/objects/MapButton.java (hunk 1 above)
Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/objects/MapButton.java]: Oh yeah I forgot to change this after I discovered .parseBoolean doesn't actually convert 1s and 0s

File: src/main/java/com/wynntils/modules/map/overlays/objects/MapTerritory.java (hunk 2 above)
Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/objects/MapTerritory.java]: No idea, I even printed both the unformatted and formatted text - they yielded the same results (the strings in the if/else right now), and neither worked with .equals() - I can try again in a bit

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java (hunk 4 above)
Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: Yeah I'll change this

File: src/main/java/com/wynntils/modules/map/overlays/objects/MapTerritory.java (hunk 2 above)
Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/map/overlays/objects/MapTerritory.java]: Use `StringUtils#stripColorCodes` from Minecraft to actually remove color codes, sometimes formatted and unformatted does not work as expected.

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java (hunk 5 above)
Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: No way to filter med+higher, we could try something like shift-click to view that diff and higher?

Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: > No way to filter med+higher, we could try something like shift-click to view that diff and higher?

Yeah, maybe that, or some kind of button for that aswell?

Not sure, leaving it up to you, just try to make it as intuitive as possible.

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java (hunk 3 above)
Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: I think we should keep this as int; it'll be more consistent with the rest of the buttons
With an enum, it'd probably make more sense to create an enum for all the defense levels in MapTerritory and reference that, but that'd prevent us from having an off state. If we included an off state in the enum, future uses on that enum would be difficult as it'd have to excluded the off state.

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java (hunk 5 above)
Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: @DonkeyBlaster Are you planning on working on this or are you done with this PR?

Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: Working, though I'm trying to fix my item highlights first

**Review by kristofbolyai — CHANGES_REQUESTED** [review]
Looks good! 

I have found two things that I find a bit strange:

- Should we disable trade routes when using the filter? 
image
Doesn't that image look a bit weird? Maybe disable all trade routes or only have them between drawn territories? I am up for discussion on this one.

- A bug I found while using the filter: Guild map has the ability to run `/gu territory <territory>` when the user Shift + Right clicks on a territory. This currently works with territories that are filtered. I would expect this to only work with territories that are drawn on the screen.

**Comment by ryanzhoudev** [comment]
I'll look into how both of these work - preferably, we'd have trade routes show for the territories that do appear.

**Review by kristofbolyai — APPROVED** [review]

File: src/main/java/com/wynntils/modules/map/overlays/objects/MapButton.java [hunk 6]
```diff
@@ -22,25 +22,27 @@
...
-    Function<Void, Boolean> isEnabled;
+    Function<Void, Integer> state;
+    Boolean isBool;
 
     ScreenRenderer renderer = new ScreenRenderer();
 
-    public MapButton(int posX, int posY, MapButtonType type, List<String> hoverLore, Function<Void, Boolean> isEnabled, BiConsumer<MapButton, Integer> onClick) {
-        int halfWidth = type.getWidth() / 2;
-        int halfHeight = type.getHeight() / 2;
+    public MapButton(int posX, int posY, MapButtonIcon icon, List<String> hoverLore, Boolean isBool, Function<Void, Integer> state, BiConsumer<MapButton, Integer> onClick) {
```
Comment (Incompleteusern) [inline src/main/java/com/wynntils/modules/map/overlays/objects/MapButton.java]: replace `Function<Void, Integer> state` with `Supplier<Integer>`

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java [hunk 7]
```diff
@@ -57,37 +68,108 @@ public void initGui() {
...
+                territoryDefenseFilterType = TDFTypes.NORMAL;
+            }
+
+            if (territoryDefenseFilter >= 5 || (territoryDefenseFilterType != TDFTypes.NORMAL && territoryDefenseFilter == 4)) { // Safety check and reset
+                territoryDefenseFilter = 0;
+            } else {
+                ++territoryDefenseFilter;
+            }
+
+            // Remove redundant settings (Very low and lower, very high and lower, very low and higher, very high and higher)
```
Comment (Incompleteusern) [inline src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java:133]: Are `very high and lower` and `very high and higher` filtered/

**Review by Incompleteusern — CHANGES_REQUESTED** [review]
a few things

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java (hunk 7 above)
Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java:133]: What do you mean by filtered? Currently it's just not possible to reach that filter state

Comment (Incompleteusern) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java:133]: wdym?

Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java:133]: Clicking through the options will skip the settings in that comment

**Review by Incompleteusern — APPROVED** [review]

File: src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java [hunk 8]
```diff
@@ -57,37 +68,108 @@ public void initGui() {
...
+                type = RESET + " and higher";
+                break;
+            case LOWER:
+                type = RESET + " and lower";
+                break;
+            case NORMAL:
+            default:
+                type = "";
+                break;
+        }
```
Comment (HighCrit) [inline src/main/java/com/wynntils/modules/map/overlays/ui/GuildWorldMapUI.java]: Personally would've opted for a method in the enum

File: src/main/java/com/wynntils/modules/map/overlays/ui/WorldMapUI.java [hunk 9]
```diff
@@ -123,15 +124,17 @@ public void initGui() {
         Keyboard.enableRepeatEvents(true);
     }
 
-    protected void addButton(MapButtonType type, int offsetX, List<String> hover, Function<Void, Boolean> isEnabled, BiConsumer<MapButton, Integer> onClick) {
+    protected MapButton addButton(MapButtonIcon icon, int offsetX, List<String> hover, Boolean isBool, Supplier<Integer> state, BiConsumer<MapButton, Integer> onClick) {
```
Comment (HighCrit) [inline src/main/java/com/wynntils/modules/map/overlays/ui/WorldMapUI.java]: Why did state become an integer if all use cases are in conditions?

Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/map/overlays/ui/WorldMapUI.java]: I don't actually remember too well - I think I had a reason to do with how I was going to inform the user of which filter state was currently selected but the current implementation doesn't need it. I'll look at reverting it back to what we had before.

**Pull request merged**